In [97]:
import numpy as np,pandas as pd,seaborn as sns

In [98]:
import sklearn

In [119]:
gbt=pd.read_csv(r'C:\Users\Sam\Downloads\groupbyExcercise.csv')
gbt

Unnamed: 0        Date Product  Revenue  Quantity
0            0  01-01-2023       A      221         3
1            1  01-01-2023       B      111         8
2            2  02-01-2023       B      171        11
3            3  02-01-2023       A      141        15
4            4  01-02-2023       A       75         4
5            5  01-02-2023       B       90        12
6            6  02-02-2023       A       96        17
7            7  02-02-2023       B      170        10
8            8  01-03-2023       B      200        19
9            9  01-03-2023       A      120        10
10          10  02-03-2023       A      121        12
11          11  02-03-2023       B      150         5

In [120]:
gbt.drop(columns=['Unnamed: 0'],inplace=True)

### When we are not passing index = False, an extra index column “Unnamed:0 “ is created

In [121]:
gbt

Date Product  Revenue  Quantity
0   01-01-2023       A      221         3
1   01-01-2023       B      111         8
2   02-01-2023       B      171        11
3   02-01-2023       A      141        15
4   01-02-2023       A       75         4
5   01-02-2023       B       90        12
6   02-02-2023       A       96        17
7   02-02-2023       B      170        10
8   01-03-2023       B      200        19
9   01-03-2023       A      120        10
10  02-03-2023       A      121        12
11  02-03-2023       B      150         5

In [122]:
gbt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Date      12 non-null     object
 1   Product   12 non-null     object
 2   Revenue   12 non-null     int64 
 3   Quantity  12 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 516.0+ bytes


In [123]:
gbt.nunique()

Date         6
Product      2
Revenue     12
Quantity    10
dtype: int64

1. Arrange DataFrame df, focusing on distinct product categories. Explore the data to reveal:
### Overall revenue insights

In [124]:
x = gbt.groupby('Product')['Revenue'].sum().reset_index()
y= x.sort_values(by='Revenue', ascending=False)
print(y)

  Product  Revenue
1       B      892
0       A      774


* the overall revenue insights is 774 for product 'A' and 892 for product 'B',by this we can observe that the revenue of product B is greater than product A in the given time period

### Average revenue patterns

In [125]:
gbt.groupby(['Product'])['Revenue'].mean()

Product
A    129.000000
B    148.666667
Name: Revenue, dtype: float64

### Peak quantity sold observations

In [126]:
gbt.groupby(['Product'])['Quantity'].max()

Product
A    17
B    19
Name: Quantity, dtype: int64

### Minimum quantity sold occurrences


In [127]:
s=gbt.groupby(['Product'])['Quantity'].min()

In [128]:
s

Product
A    3
B    5
Name: Quantity, dtype: int64

2. Aggregate DataFrame df by extracting the month and year from the 'Date' column. Calculate the collective revenue for each unique combination of month and year.

### The pd.to_datetime function in pandas is used to convert argument to datetime. In the context of a DataFrame, it's often used to convert a column containing date-like strings or other date-like objects into a pandas datetime object, which allows for convenient datetime operations.

In [137]:
gbt['Date'] = pd.to_datetime(gbt['Date'])
gbt['Month'] = gbt['Date'].dt.month
gbt['Year'] = gbt['Date'].dt.year
gbt[['Date','Month','Year']]

Date  Month  Year
0  2023-01-01      1  2023
1  2023-01-01      1  2023
2  2023-02-01      2  2023
3  2023-02-01      2  2023
4  2023-01-02      1  2023
5  2023-01-02      1  2023
6  2023-02-02      2  2023
7  2023-02-02      2  2023
8  2023-01-03      1  2023
9  2023-01-03      1  2023
10 2023-02-03      2  2023
11 2023-02-03      2  2023

In [143]:
x = gbt.groupby(['Month', 'Year'])['Revenue'].sum()
print(x)

Month  Year
1      2023    817
2      2023    849
Name: Revenue, dtype: int64


In [158]:
gbt['Date'] = pd.to_datetime(gbt['Date'])
gbt['MonthYear'] = gbt['Date'].dt.to_period("M")
revenue_by_month_year = gbt.groupby('MonthYear')['Revenue'].sum().reset_index()
print(revenue_by_month_year)

  MonthYear  Revenue
0   2023-01      817
1   2023-02      849


3. Group the DataFrame df by the 'Date' column and find the date with the highest total revenue and the date with the highest total quantity sold.


In [159]:
hr= gbt.groupby('Date').agg({'Revenue': 'sum', 'Quantity': 'sum'}).reset_index()
hr

Date  Revenue  Quantity
0 2023-01-01      332        11
1 2023-01-02      165        16
2 2023-01-03      320        29
3 2023-02-01      312        26
4 2023-02-02      266        27
5 2023-02-03      271        17

### reset_index :
* the reset_index method in pandas is used to reset the index of a DataFrame. When you perform operations that result in a new DataFrame, the index might not be in the default integer range (0, 1, 2, ...). Using reset_index helps in bringing the DataFrame back to a default integer index and turns the previous index into a regular column.

In [160]:
max_revenue_date = hr.loc[grouped_by_date['Revenue'].idxmax()]['Date']
print(f"Date with highest total revenue: {max_revenue_date}")

Date with highest total revenue: 2023-01-01 00:00:00


### idxmax():
* In pandas, the idxmax() function is used to get the index of the first occurrence of the maximum value in a Series.
*  idxmax() is used to find the index of the maximum value in the 'Revenue' and 'Quantity' columns within the grouped DataFrame (hr). Then, loc is used to locate the corresponding 'Date' value for that index.


In [162]:
max_quantity_date = hr.loc[grouped_by_date['Quantity'].idxmax()]['Date']
print(f"Date with highest total quantity sold: {max_quantity_date}")

Date with highest total quantity sold: 2023-01-03 00:00:00


4.  Group the Data Frame df by the Date and find products that have a total quantity sold greater than 10 and less than 15 .


In [164]:
x = gbt.groupby(['Date', 'Product'])['Quantity'].sum()

In [165]:
y = x[(x > 10) & (x < 15)]

In [166]:
y

Date        Product
2023-01-02  B          12
2023-02-01  B          11
2023-02-03  A          12
Name: Quantity, dtype: int64

5. Group the DataFrame df by the 'Product' column and
   add a new column 'Revenue Percentage' that shows the percentage of each product's revenue relative to the total revenue for that product.



In [168]:
x = gbt.groupby('Product')['Revenue'].sum()

In [169]:
percentage = x / x.sum() * 100

In [171]:
gbt['Revenue Percentage'] = gbt['Product'].map(percentage)
print(gbt)

         Date Product  Revenue  Quantity Month_Year  Month  Year MonthYear  \
0  2023-01-01       A      221         3    2023-01      1  2023   2023-01   
1  2023-01-01       B      111         8    2023-01      1  2023   2023-01   
2  2023-02-01       B      171        11    2023-02      2  2023   2023-02   
3  2023-02-01       A      141        15    2023-02      2  2023   2023-02   
4  2023-01-02       A       75         4    2023-01      1  2023   2023-01   
5  2023-01-02       B       90        12    2023-01      1  2023   2023-01   
6  2023-02-02       A       96        17    2023-02      2  2023   2023-02   
7  2023-02-02       B      170        10    2023-02      2  2023   2023-02   
8  2023-01-03       B      200        19    2023-01      1  2023   2023-01   
9  2023-01-03       A      120        10    2023-01      1  2023   2023-01   
10 2023-02-03       A      121        12    2023-02      2  2023   2023-02   
11 2023-02-03       B      150         5    2023-02      2  2023